compiler/rustc_mir/src/borrow_check/type_check/mod.rs

```diff
@@ -1273,6 +1273,17 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
 
         // the "concrete opaque types" maps
         let concrete_opaque_types = &tcx.typeck(anon_owner_def_id).concrete_opaque_types;
+        debug!("concrete_opaque_types from typeck: {:#?}", concrete_opaque_types);
+        let concrete_opaque_types: VecMap<_, _> = concrete_opaque_types
+            .iter()
+            .map(|(k, v)| {
+                let k =
+                    self.borrowck_context.universal_regions.indices.fold_to_region_vids(tcx, *k);
```

Comment:
My current status is that this introduces an ICE, because the lifetimes in the key (tait def id + substs) has the substs from some intermediate type (in the test mentioned later, this would be the impl block with the type alias that is *neither a declaration nor a defining use*), and hasn't gotten them replaced with the ones from the local function. This means that the `DefId` of the lifetime will differ. You can observe this in `issue-70121.rs`. It becomes more visible in debug logs if you change the lifetime names, so not everything is `'a` but has a unique name.

Reply:
```rust
// declaration
pub type Successors<'a> = impl Iterator<Item = &'a ()>;

// first defining use, ok
pub fn f<'a>() -> Successors<'a> {
    None.into_iter()
}

pub trait Tr {
    type Item;
}

impl<'a> Tr for &'a () {
    // not a defining use
    type Item = Successors<'a>;
}

// second defining use, ICEs, because the `DefId` + `Substs` are "Successors", `'a` (but a different `'a` than the one on `kazusa`)
pub fn kazusa<'a>() -> <&'a () as Tr>::Item {
    None.into_iter()
}
```

Reply:
somewhere in https://github.com/rust-lang/rust/blob/b9197978a90be6f7570741eabe2da175fec75375/compiler/rustc_typeck/src/check/mod.rs#L505-L513 things go wrong if the opaque type was only mentioned after normalization

Reply:
oh yea, this is wrong

https://github.com/rust-lang/rust/blob/8240e7aa101815e2009c7d03b33dd2566d843e73/compiler/rustc_typeck/src/check/mod.rs#L381

the defid is of the tait, not the function, so we're inserting the wrong parameters

Reply:
`fixup_opaque_types` is really suspicious, I would definitely throw it out an airlock on a whim.

I understand what it does, it works in the documented example (TAIT in impl block's associated type), but I don't think it works anywhere else and is likely the source of a few weird ICEs.

Reply:
I believe this entire new code became obsolete with https://github.com/rust-lang/rust/pull/87200